### ISSUE TITLE
Empty lines in @file/@clean trees

### ISSUE THREAD

**Issue by jurov** [leo-editor/leo-editor#248]
If you put some empty line(s) in the beginning and/or the end of nodes in @clean tree, they are not consistently output into the generated file. With sample [workbook.leo.txt](https://github.com/leo-editor/leo-editor/files/195591/workbook.leo.txt) , I am getting:
```
line1


line5

line8
```
If there is some stripping going on, is it documented anywhere?

**Comment by tbnorth** [comment]
Just a note - I thought this might be a duplicate, but failed to find another open bug on this issue.

**Comment by edreamleo** [comment]
Everything works for me as expected, with my settings, on your test file.

As with @shadow, there may be glitches when updating the outline from changes made externally. These changes will "take" when you save the .leo file. As I have just verified, Leo's "Recovered Node" logic shows where all such external changes happen.  You must then verify that the "boundary lines" are where they should be, moving them as needed.

I see no @clean-related settings, and I see no special cases for @clean in the read/write code:

    at.readOneAtCleanNode # @shadow-like logic
    at.write
    at.writeAllHelper
    at.writeOpen file

In short, everything appears to be working as expected. There was a proposal to adjust empty lines automatically when reading an @clean file, but that's a separate, low-priority, issue, apparently not in the bug tracker.

**Comment by edreamleo** [comment]
I am going to close this.  Please reopen if you can demonstrate the bug more plausibly.

**Comment by jurov** [comment]
I have no idea what was different on your side, please specify what "works as expected" and "plausible demonstration" mean? Namely, did you get same output file I quoted, or something else? And is that one correct or not?

I now retried it on a computer where Leo was not installed,
1. cloned git repository and ran it from master branch. 
2. downloaded the sample workbook.leo.txt, renamed to workbook.leo and opened it 
3. verified that the clean tree has 3+3+2 = 8 lines total (as shown by Leo's own line numbering). 
4. Menu File> Read-Write > Write @file Nodes

Now the file /var/tmp/tmp2.txt has 6 lines, and it should have 8 lines. This difference is IMO a bug. There was no updating from disk whatsoever.

**Comment by edreamleo** [comment]
[Here](https://github.com/benzolius/leo-scripts/blob/master/correct_empty_lines.py) is the script I mentioned above.

**Comment by jurov** [comment]
But again, there was no updating from disk whatsoever, I don't get what is there to "correct". I created the clean tree in Leo!

**Comment by edreamleo** [comment]
@jurov I'll reopen this.  Didn't mean to upset you :-) I do think that this is some kind of misunderstanding.

**Comment by jurov** [comment]
Thanks. Actually, reading back is funny too, if I add empty lines into the /var/tmp/tmp2.txt file so that line5 and line8 are really 5th and 8th, and read it back, there are superfluous newlines in @clean tree.

**Comment by edreamleo** [comment]
@jurov How are we going to resolve this? We could chat on the phone or on irc.

**Comment by edreamleo** [comment]
@jurov Have you tried clearing caches?  Sometimes the cache interferes with updates, leading to bizarre behavior.

I do not believe there is anything wrong with @clean.  In the last several weeks I have created half a dozen or so @clean test files, and they have always performed as expected.  There are also unit tests for @clean. I have used @clean successfully in several external project. Furthermore, I would expect many other people to complain if there were an actual bug in @clean.

I am going to close this again.  I suggest asking for help on leo-editor.  That discussion will have a larger audience, and may produce new ideas.
